index.js

```diff
@@ -1,171 +1,126 @@
+        this.gridInner = new Float64Array(gridSize * gridSize);
+        this.f = new Float64Array(gridSize);
+        this.z = new Float64Array(gridSize + 1);
+        this.v = new Uint16Array(gridSize);
+    }
 
-    this.ctx.textAlign = 'left'; // Necessary so that RTL text doesn't have different alignment
-    this.ctx.fillStyle = 'black';
+    _createCanvas(size) {
+        const canvas = document.createElement('canvas');
+        canvas.width = canvas.height = size;
+        return canvas;
+    }
 
-    // temporary arrays for the distance transform
-    this.gridOuter = new Float64Array(gridSize * gridSize);
-    this.gridInner = new Float64Array(gridSize * gridSize);
-    this.f = new Float64Array(gridSize);
-    this.z = new Float64Array(gridSize + 1);
-    this.v = new Uint16Array(gridSize);
+    getMetrics(char) {
+        const {
+            width: advance,
+            actualBoundingBoxAscent,
+            actualBoundingBoxDescent,
+            actualBoundingBoxLeft,
+            actualBoundingBoxRight
+        } = this.ctx.measureText(char);
 
-    this.useMetrics = this.ctx.measureText('A').actualBoundingBoxLeft !== undefined;
+        // The integer/pixel part of the top alignment is encoded in metrics.top
+        // The remainder is implicitly encoded in the rasterization
+        const top = Math.floor(actualBoundingBoxAscent);
+        const left = 0;
 
-    // hack around https://bugzilla.mozilla.org/show_bug.cgi?id=737852
-    this.middle = Math.round((size / 2) * (navigator.userAgent.indexOf('Gecko/') >= 0 ? 1.2 : 1));
-}
+        // If the glyph overflows the canvas size, it will be clipped at the bottom/right
+        const width = Math.min(this.size, Math.ceil(actualBoundingBoxRight - actualBoundingBoxLeft));
+        const height = Math.min(this.size - this.buffer, Math.ceil(actualBoundingBoxAscent + actualBoundingBoxDescent));
 
-function prepareGrids(imgData, width, height, glyphWidth, glyphHeight, gridOuter, gridInner) {
-    // Initialize grids outside the glyph range to alpha 0
-    gridOuter.fill(INF, 0, width * height);
-    gridInner.fill(0, 0, width * height);
-
-    var offset = (width - glyphWidth) / 2; // This is zero if we're not extracting metrics
-
-    for (var y = 0; y < glyphHeight; y++) {
-        for (var x = 0; x < glyphWidth; x++) {
-            var j = (y + offset) * width + x + offset;
-            var a = imgData.data[4 * (y * glyphWidth + x) + 3] / 255; // alpha value
-            if (a === 1) {
-                gridOuter[j] = 0;
-                gridInner[j] = INF;
-            } else if (a === 0) {
-                gridOuter[j] = INF;
-                gridInner[j] = 0;
-            } else {
-                var b = Math.max(0, 0.5 - a);
-                var c = Math.max(0, a - 0.5);
-                gridOuter[j] = b * b;
-                gridInner[j] = c * c;
-            }
-        }
-    }
-}
+        const sdfWidth = width + 2 * this.buffer;
```

Comment:
It feels a little clunky that we return both a "width" and an "sdfWidth" that are just different by a constant. There were two reasons I introduced it:

- At first, `sdfWidth` and `width` could be different in some clipping cases (this is no longer true).
- It's weird to make the caller responsible for knowing the right conversion from width -> sdfWidth. (e.g. during development I would accidentally mismatch them and find out when `AlphaImage.copy` threw an exception).

The first concern isn't an issue any more. I wonder if there's a better way to handle the second. One option would be to just use the SDF/bitmap width for the metrics as well. The only place I know it's used in GL JS is to do center-alignment for vertical labels, which would still work with a constant offset but would be shifted a pixel or two. If TinySDF were part of GL JS, I'd just say "it should return an `AlphaImage`", but we probably don't want that dependency...